File: Helpers/src/main/java/com/dg/helpers/SchedulerHandler.java
package com.dg.helpers;

import android.os.Handler;
import android.os.Looper;

import java.util.HashMap;

public class SchedulerHandler extends Handler
{
    private HashMap<Runnable, Runnable> mTasks = new HashMap<>();

    public SchedulerHandler()
    {
        super();
    }

    public SchedulerHandler(Callback callback)
    {
        super(callback);
    }

    public SchedulerHandler(Looper looper)
    {
        super(looper);
    }

    public SchedulerHandler(Looper looper, Callback callback)
    {
        super(looper, callback);
    }

    public void postWithSchedule(final Runnable task, long delay, final long period)
    {
        Runnable runnable = new Runnable()
        {
            @Override
            public void run()
            {
                task.run();
                postDelayed(this, period);
            }
        };

        mTasks.put(task, runnable);

        postDelayed(runnable, delay);
    }

    public void postWithSchedule(final Runnable task, final long period)
    {
        postWithSchedule(task, period, period);
    }

    public void removeScheduledTask(Runnable task)
    {
        Runnable removed = mTasks.remove(task);
        if (removed == null) return;

        removeCallbacks(removed);
    }
}

File: Helpers/src/main/java/com/dg/helpers/SchedulerHandler.kt
package com.dg.helpers

import android.os.Handler
import android.os.Looper

import java.util.HashMap

@Suppress("unused")
class SchedulerHandler : Handler
{
    private val mTasks = HashMap<Runnable, Runnable>()

    constructor() : super()

    constructor(callback: Callback) : super(callback)

    constructor(looper: Looper) : super(looper)

    constructor(looper: Looper, callback: Callback) : super(looper, callback)

    @Suppress("MemberVisibilityCanBePrivate")
    fun postWithSchedule(task: Runnable, delay: Long, period: Long)
    {
        val runnable = object : Runnable
        {
            override fun run()
            {
                task.run()
                postDelayed(this, period)
            }
        }

        mTasks[task] = runnable

        postDelayed(runnable, delay)
    }

    fun postWithSchedule(task: Runnable, period: Long)
    {
        postWithSchedule(task, period, period)
    }

    fun removeScheduledTask(task: Runnable)
    {
        val removed = mTasks.remove(task) ?: return

        removeCallbacks(removed)
    }
}
